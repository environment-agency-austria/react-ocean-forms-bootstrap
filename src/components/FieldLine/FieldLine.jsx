/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Label, Col, InputGroup } from 'reactstrap';
import { FormText } from 'react-ocean-forms';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms-legacy';

import { FieldRow } from './FieldRow';
import { RequiredMarker } from './RequiredMarker';
import { InvalidAlert } from './InvalidAlert';
import { ValidatingSpinner } from './ValidatingSpinner';
import { FieldLineAddon } from './FieldLineAddon';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import FieldError from './FieldError';

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
class FieldLine extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      infoVisible: false,
    };

    this.toggleInfo = this.toggleInfo.bind(this);
  }

  /**
   * Toggles the visibility of the info alert
   */
  toggleInfo() {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  render() {
    const {
      field,
      meta,
      label,
      children,
      validators,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} validators={validators} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

FieldLine.displayName = 'FieldLine';

FieldLine.defaultProps = {
  info: undefined,
  validators: undefined,
  prefix: undefined,
  suffix: undefined,
  labelSize: 3,
  inputSize: 9,
  labelClass: 'text-right',
};

FieldLine.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
  prefix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  suffix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  labelSize: PropTypes.string,
  inputSize: PropTypes.string,
  labelClass: PropTypes.string,
};

export default FieldLine;
