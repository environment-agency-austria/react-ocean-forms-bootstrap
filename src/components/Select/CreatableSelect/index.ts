export * from './CreatableSelect';
export * from './CreatableSelect.types';
