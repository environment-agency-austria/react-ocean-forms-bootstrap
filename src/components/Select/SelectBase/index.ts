export * from './SelectBase';
export * from './SelectBase.types';
