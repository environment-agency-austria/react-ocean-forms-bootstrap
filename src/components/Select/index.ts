export * from './SelectBase';
export * from './Select';
export * from './CreatableSelect';
