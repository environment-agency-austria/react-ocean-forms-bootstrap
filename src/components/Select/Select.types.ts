import { IFieldComponentProps } from 'react-ocean-forms';
import { IInfoProps } from '../FieldLine';

/**
 * Selectable option
 */
export interface ISelectOption {
  /**
   * Display label
   */
  label: string;
  /**
   * Internal value
   */
  value: string;
}

// tslint:disable-next-line:no-any
export function isSelectOption(object: any): object is ISelectOption {
  return object && typeof object.label === 'string' && typeof object.value === 'string';
}

/**
 * Type for an array of ISelectOption
 * @see ISelectOption
 */
export type ISelectOptions = ISelectOption[];

/**
 * Props for the Select component
 */
export interface ISelectProps extends IFieldComponentProps, IInfoProps {
  /**
   * Placeholder string that will be displayed while
   * the component is empty
   */
  placeholder: string;
  /**
   * True, if the select component should be in multi-select mode
   */
  multi: boolean;
  /**
   * Selectable options
   */
  options: ISelectOptions;
}
