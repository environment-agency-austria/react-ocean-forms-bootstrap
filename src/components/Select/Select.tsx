/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { withField } from 'react-ocean-forms';
import { default as ReactSelect } from 'react-select';
import { Input as StrapInput } from 'reactstrap';

import { FieldLine } from '../FieldLine';
import { ISelectOption, ISelectProps, isSelectOption } from './Select.types';

/**
 * Component for displaying bootstrap
 * form groups with an select input and
 * oForm support
 */
export class BaseSelect extends React.Component<ISelectProps> {
  public static displayName: string = 'Select';

  // tslint:disable-next-line:typedef
  public static defaultProps = {
    multi: false,
    placeholder: 'ojs_select_placeholder',
  };

  /**
   * Manually handle the onChange event of
   * the select because oForms is expecting
   * default input onChange behaviour.
   */
  private handleChange = (value: ISelectOption): void => {
    const { field } = this.props;

    field.onChange({
      target: {
        value,
      },
    });
  }

  /**
   * Manually handle the onBlur event of
   * the select because oForms is expecting
   * default input onBlur behaviour.
   */
  private handleBlur = (): void => {
    const { field } = this.props;

    field.onBlur();
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      field,
      placeholder,
      options,
      multi,
      meta,
    } = this.props;

    // Generate a css class based on the validity of the select element
    let selectClass = 'react-select-control';
    selectClass = (selectClass + (meta.valid ? '' : ' is-invalid')).trim();

    // Check if the current value has a different label than the value
    // with the same key in the options array. Bugfix to change the
    // selected label when the current language changes.
    let fieldValue: ISelectOption | undefined;
    if (isSelectOption(field.value)) {
      fieldValue = field.value;
      // tslint:disable-next-line:no-non-null-assertion
      const selectableValue = options.find(item => item.value === fieldValue!.value);
      if (selectableValue !== undefined && selectableValue.label !== field.value.label) {
        field.value.label = selectableValue.label;
        this.handleChange(field.value);
      }
    }

    // Support for plaintext display
    if (meta.plaintext) {
      let displayValue = '';
      if (isSelectOption(field.value)) {
        displayValue = field.value.label;
      }

      return (
        <FieldLine {...this.props}>
          <StrapInput {...field} value="" plaintext>
            {displayValue}
          </StrapInput>
        </FieldLine>
      );
    }

    const placeholderText = meta.stringFormatter(placeholder);

    return (
      <FieldLine {...this.props}>
        <ReactSelect
          id={field.id}
          value={fieldValue}
          isDisabled={field.disabled}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
          options={options}
          isMulti={multi}
          placeholder={placeholderText}
          className={selectClass}
        />
      </FieldLine>
    );
  }
}

export const Select = withField(BaseSelect);
