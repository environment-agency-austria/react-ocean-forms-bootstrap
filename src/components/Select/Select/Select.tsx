/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { default as ReactSelect } from 'react-select';

import { IPreparedSelectProps, SelectBase, ISelectFieldValue } from '../SelectBase';
import { ISelectProps } from './Select.types';

/**
 * Renders the actual `Select` component with the passed props
 */
const renderSelect = (preparedProps: IPreparedSelectProps): JSX.Element => {
  return (
    <ReactSelect {...preparedProps} />
  );
}

/**
 * Component for displaying bootstrap
 * form groups with an select input and
 * oForm support
 */
export const Select = <TSubmitValue extends unknown = ISelectFieldValue>(props: ISelectProps<TSubmitValue>): JSX.Element => {
  const {
    ...selectBaseProps
  } = props;

  return (
    <SelectBase
      {...selectBaseProps}
      renderSelect={renderSelect}
    />
  );
};
