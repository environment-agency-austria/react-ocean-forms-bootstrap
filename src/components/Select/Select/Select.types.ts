/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Subtract } from 'react-ocean-forms';

import { ISelectBasePropsBase, SelectBase } from '../SelectBase';

type ISelectBaseDefaultProps = (typeof SelectBase)['defaultProps'];
type WithExcluded = Subtract<ISelectBasePropsBase, ISelectBaseDefaultProps>;

/**
 * Props for the component `Select`
 */
export interface ISelectProps extends WithExcluded, Partial<ISelectBaseDefaultProps> {
}
