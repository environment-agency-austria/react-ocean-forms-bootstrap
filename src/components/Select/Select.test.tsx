import * as React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { IFieldComponentFieldProps, IFieldComponentMeta } from 'react-ocean-forms';
import { default as ReactSelect } from 'react-select';

import { createMockField, createMockFieldMeta } from '../../test-utils/enzymeFormContext';
import { FieldLine } from '../FieldLine';
import { BaseSelect } from './Select';
import { ISelectOption, ISelectOptions, ISelectProps } from './Select.types';

describe('<Select />', () => {
  interface ISetupArgs {
    props?: Partial<ISelectProps>;
    fieldOverrides?: Partial<IFieldComponentFieldProps>;
    metaOverrides?: Partial<IFieldComponentMeta>;
  }

  interface ISetupResult {
    wrapper: ShallowWrapper;
    field: IFieldComponentFieldProps;
    meta: IFieldComponentMeta;
    options: ISelectOptions;
  }

  const setup = ({
    props,
    metaOverrides,
    fieldOverrides,
  }: ISetupArgs = {}): ISetupResult => {
    const fieldLabel = 'field0';
    const meta = {
      ...createMockFieldMeta(),
      ...metaOverrides,
    };
    const field = {
      ...createMockField(),
      value: '',
      ...fieldOverrides,
    };

    const options = [
      { value: 'one', label: 'One' },
      { value: 'two', label: 'Two' },
    ];

    const wrapper = shallow((
      <BaseSelect
        label={fieldLabel}
        meta={meta}
        field={field}
        options={options}
        {...props}
      />
    ));

    return {
      wrapper,
      field,
      meta,
      options,
    };
  };

  it('should render correctly', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a field line and a react-select', () => {
    const { wrapper } = setup();

    expect(wrapper.find(FieldLine).exists()).toBeTruthy();
    expect(wrapper.find(ReactSelect).exists()).toBeTruthy();
  });

  it('should pass the meta props to the field line', () => {
    const { wrapper, meta } = setup();

    expect(wrapper.find(FieldLine).prop('meta')).toEqual(meta);
  });

  describe('label edge case', () => {
    it('should update the label of the selected value if it does not match with the label of the item inside options', () => {
      const { field } = setup({
        fieldOverrides: { value: { value: 'two', label: '' } },
      });

      const fieldValue: ISelectOption = field.value as ISelectOption;
      expect(fieldValue.label).toBe('Two');
    });

    it('should trigger the field.onChange event if the labels mismatch', () => {
      const { field } = setup({
        fieldOverrides: { value: { value: 'two', label: '' } },
      });

      expect(field.onChange).toHaveBeenCalledWith({
        target: {
          value: { value: 'two', label: 'Two' },
        },
      });
    });
  });

  describe('field events', () => {
    it('should call field.onChange when the input changes', () => {
      const { wrapper, field, options } = setup();

      const changeProp = wrapper.find(ReactSelect).prop('onChange');
      changeProp(options[1]);

      expect(field.onChange).toHaveBeenCalledWith({
        target: {
          value: options[1],
        },
      });
    });

    it('should call field.onBlur when there is an input blur', () => {
      const { wrapper, field } = setup();

      const blurProp = wrapper.find(ReactSelect).prop('onBlur');
      blurProp();

      expect(field.onBlur).toHaveBeenCalledWith();
    });
  });

  describe('meta.plaintext', () => {
    it('should correctly react to meta.plaintext', () => {
      const { wrapper } = setup({
        metaOverrides: { plaintext: true },
      });
      expect(wrapper).toMatchSnapshot();
    });

    it('should display the label property of the selected option in plaintext mode', () => {
      const { wrapper } = setup({
        metaOverrides: { plaintext: true },
        fieldOverrides: { value: { value: 'two', label: 'Two' } },
      });
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should set the is-invalid css class to the ReactSelect component if meta.valid is false', () => {
    const { wrapper } = setup({
      metaOverrides: { valid: false },
    });
    expect(wrapper.find(ReactSelect).prop('className')).toBe('react-select-control is-invalid');
  });
});
