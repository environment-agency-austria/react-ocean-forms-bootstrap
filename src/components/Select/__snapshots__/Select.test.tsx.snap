// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> meta.plaintext should correctly react to meta.plaintext 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    field={
      Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="field0"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    multi={false}
    options={
      Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ]
    }
    placeholder="ojs_select_placeholder"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Input
        disabled={false}
        id="mock-field"
        name="mock-field"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        plaintext={true}
        type="text"
        value=""
      >
        
      </Input>,
      "field": Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      },
      "inputSize": "9",
      "label": "field0",
      "labelClass": "text-right",
      "labelSize": "3",
      "meta": Object {
        "error": null,
        "isValidating": false,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      },
      "multi": false,
      "options": Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ],
      "placeholder": "ojs_select_placeholder",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "",
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "plaintext": true,
        "type": "text",
        "value": "",
      },
      "ref": null,
      "rendered": "",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Input
          disabled={false}
          id="mock-field"
          name="mock-field"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          plaintext={true}
          type="text"
          value=""
        >
          
        </Input>,
        "field": Object {
          "disabled": false,
          "id": "mock-field",
          "name": "mock-field",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "",
        },
        "inputSize": "9",
        "label": "field0",
        "labelClass": "text-right",
        "labelSize": "3",
        "meta": Object {
          "error": null,
          "isValidating": false,
          "plaintext": true,
          "stringFormatter": [MockFunction],
          "touched": false,
          "valid": true,
        },
        "multi": false,
        "options": Array [
          Object {
            "label": "One",
            "value": "one",
          },
          Object {
            "label": "Two",
            "value": "two",
          },
        ],
        "placeholder": "ojs_select_placeholder",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "",
          "disabled": false,
          "id": "mock-field",
          "name": "mock-field",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "plaintext": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": "",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<Select /> meta.plaintext should display the label property of the selected option in plaintext mode 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    field={
      Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": Object {
          "label": "Two",
          "value": "two",
        },
      }
    }
    label="field0"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    multi={false}
    options={
      Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ]
    }
    placeholder="ojs_select_placeholder"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Input
        disabled={false}
        id="mock-field"
        name="mock-field"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        plaintext={true}
        type="text"
        value=""
      >
        Two
      </Input>,
      "field": Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": Object {
          "label": "Two",
          "value": "two",
        },
      },
      "inputSize": "9",
      "label": "field0",
      "labelClass": "text-right",
      "labelSize": "3",
      "meta": Object {
        "error": null,
        "isValidating": false,
        "plaintext": true,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      },
      "multi": false,
      "options": Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ],
      "placeholder": "ojs_select_placeholder",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "Two",
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "plaintext": true,
        "type": "text",
        "value": "",
      },
      "ref": null,
      "rendered": "Two",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Input
          disabled={false}
          id="mock-field"
          name="mock-field"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          plaintext={true}
          type="text"
          value=""
        >
          Two
        </Input>,
        "field": Object {
          "disabled": false,
          "id": "mock-field",
          "name": "mock-field",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": Object {
            "label": "Two",
            "value": "two",
          },
        },
        "inputSize": "9",
        "label": "field0",
        "labelClass": "text-right",
        "labelSize": "3",
        "meta": Object {
          "error": null,
          "isValidating": false,
          "plaintext": true,
          "stringFormatter": [MockFunction],
          "touched": false,
          "valid": true,
        },
        "multi": false,
        "options": Array [
          Object {
            "label": "One",
            "value": "one",
          },
          Object {
            "label": "Two",
            "value": "two",
          },
        ],
        "placeholder": "ojs_select_placeholder",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Two",
          "disabled": false,
          "id": "mock-field",
          "name": "mock-field",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "plaintext": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": "Two",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<Select /> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    field={
      Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="field0"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction] {
          "calls": Array [
            Array [
              "ojs_select_placeholder",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": "ojs_select_placeholder",
            },
          ],
        },
        "touched": false,
        "valid": true,
      }
    }
    multi={false}
    options={
      Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ]
    }
    placeholder="ojs_select_placeholder"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <StateManager
        className="react-select-control"
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        id="mock-field"
        isDisabled={false}
        isMulti={false}
        onBlur={[Function]}
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "One",
              "value": "one",
            },
            Object {
              "label": "Two",
              "value": "two",
            },
          ]
        }
        placeholder="ojs_select_placeholder"
      />,
      "field": Object {
        "disabled": false,
        "id": "mock-field",
        "name": "mock-field",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      },
      "inputSize": "9",
      "label": "field0",
      "labelClass": "text-right",
      "labelSize": "3",
      "meta": Object {
        "error": null,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction] {
          "calls": Array [
            Array [
              "ojs_select_placeholder",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": "ojs_select_placeholder",
            },
          ],
        },
        "touched": false,
        "valid": true,
      },
      "multi": false,
      "options": Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ],
      "placeholder": "ojs_select_placeholder",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "className": "react-select-control",
        "defaultInputValue": "",
        "defaultMenuIsOpen": false,
        "defaultValue": null,
        "id": "mock-field",
        "isDisabled": false,
        "isMulti": false,
        "onBlur": [Function],
        "onChange": [Function],
        "options": Array [
          Object {
            "label": "One",
            "value": "one",
          },
          Object {
            "label": "Two",
            "value": "two",
          },
        ],
        "placeholder": "ojs_select_placeholder",
        "value": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <StateManager
          className="react-select-control"
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          id="mock-field"
          isDisabled={false}
          isMulti={false}
          onBlur={[Function]}
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "One",
                "value": "one",
              },
              Object {
                "label": "Two",
                "value": "two",
              },
            ]
          }
          placeholder="ojs_select_placeholder"
        />,
        "field": Object {
          "disabled": false,
          "id": "mock-field",
          "name": "mock-field",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "",
        },
        "inputSize": "9",
        "label": "field0",
        "labelClass": "text-right",
        "labelSize": "3",
        "meta": Object {
          "error": null,
          "isValidating": false,
          "plaintext": false,
          "stringFormatter": [MockFunction] {
            "calls": Array [
              Array [
                "ojs_select_placeholder",
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": "ojs_select_placeholder",
              },
            ],
          },
          "touched": false,
          "valid": true,
        },
        "multi": false,
        "options": Array [
          Object {
            "label": "One",
            "value": "one",
          },
          Object {
            "label": "Two",
            "value": "two",
          },
        ],
        "placeholder": "ojs_select_placeholder",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "react-select-control",
          "defaultInputValue": "",
          "defaultMenuIsOpen": false,
          "defaultValue": null,
          "id": "mock-field",
          "isDisabled": false,
          "isMulti": false,
          "onBlur": [Function],
          "onChange": [Function],
          "options": Array [
            Object {
              "label": "One",
              "value": "one",
            },
            Object {
              "label": "Two",
              "value": "two",
            },
          ],
          "placeholder": "ojs_select_placeholder",
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
